presto-hive/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java

```diff
@@ -0,0 +1,703 @@
+        return Optional.of(buildPartitionNames(table.getPartitionColumns(), partitions));
+    }
+
+    /**
+     * @param parts List of partition values to filter on which can be full or partial.
+     * ex: partition keys = ['a', 'b', 'c']
+     * parts could be:
+     * ['1','2','3'] or
+     * ['', '2', '']
+     * @return a list of partition names.
+     */
+    @Override
+    public Optional<List<String>> getPartitionNamesByParts(String databaseName, String tableName, List<String> parts)
+    {
+        Table table = getTableOrElseThrow(databaseName, tableName);
+        String expression = GlueExpressionUtil.buildExpression(table.getPartitionColumns(), parts);
+        List<Partition> partitions = getPartitions(databaseName, tableName, expression);
+        return Optional.of(buildPartitionNames(table.getPartitionColumns(), partitions));
+    }
+
+    private List<Partition> getPartitions(String databaseName, String tableName, String expression)
+    {
+        try {
+            List<Partition> partitions = new ArrayList<>();
+            String nextToken = null;
+
+            do {
+                GetPartitionsResult result = glueClient.getPartitions(new GetPartitionsRequest()
+                        .withDatabaseName(databaseName)
+                        .withTableName(tableName)
+                        .withExpression(expression)
+                        .withNextToken(nextToken));
+                result.getPartitions().stream()
+                        .forEach(p -> partitions.add(GlueToPrestoConverter.convertPartition(p)));
+                nextToken = result.getNextToken();
+            }
+            while (nextToken != null);
+
+            return partitions;
+        }
+        catch (AmazonServiceException e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, e);
+        }
+    }
+
+    private List<String> buildPartitionNames(List<Column> partitionColumns, List<Partition> partitions)
+    {
+        List<String> partitionNames = new ArrayList<>();
+        List<String> partitionKeys = partitionColumns.stream().map(Column::getName).collect(toList());
+        partitions.stream().forEach(p -> partitionNames.add(toPartitionName(partitionKeys, p.getValues())));
+        return partitionNames;
+    }
+
+    /**
+     * @param partitionNames is a list of full partition names
+     * ex: partition keys = ['a', 'b']
+     * partition names:
+     * ['a=1/b=2', 'a=2/b=2']
+     * @return is a mapping of partition name to partition object.
+     */
+    @Override
+    public Map<String, Optional<Partition>> getPartitionsByNames(String databaseName, String tableName, List<String> partitionNames)
+    {
+        requireNonNull(partitionNames, "partitionNames is null");
+        if (partitionNames.isEmpty()) {
+            return ImmutableMap.of();
+        }
+
+        List<Partition> partitions = batchGetPartition(databaseName, tableName, partitionNames);
```

Comment:
Why is this not using `getPartitions()`? They seem to do the same thing.

Reply:
```getPartitions()``` is intended to retrieve all partitions or used with a filter expression.  ```batchGetPartition``` needs a list of full partition names (we know exactly the list of partitions we want) It seem better suited for this method where we are given a list of specific partitions.

If possible ```getPartition``` or ```batchGetPartition``` should be preferred over ```getPartitions```. Under the hood, Glue uses DynamoDb tables and I believe ```getPartitions``` translates to a ddb ```Scan/Query``` while ```batchGetPartition``` translates to a ddb ```batchGetItem```. I will have to confirm this with Glue team. But based on this, ```getPartitions``` (with filter) would typically be more expensive than a ```batchGetPartition```

Reply:
Thanks for the explanation. That makes sense.